libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,955 @@
+        }
+   }
+
+   return !vtx_states_queue.is_empty();
+}
+
+// utility method to get power in dbm mapping to power levels
+uint8_t AP_SmartAudio::_get_power_in_dbm_from_vtx_power_level(uint8_t power_level, uint8_t& protocol_version)
+{
+    uint8_t power_in_dbm=0;
+    return _get_power_in_dbm_from_vtx_power_level(power_level,protocol_version,power_in_dbm);
+}
+// returns the dbs associated by power-level input
+uint8_t AP_SmartAudio::_get_power_in_dbm_from_vtx_power_level(uint8_t power_level, uint8_t& protocol_version, uint8_t& power_in_dbm)
+{
+
+    for (uint8_t j = 0; j < 4; j++) {
+        if (POWER_LEVELS[protocol_version][j] == power_level) {
+            switch (j)
+            {
+            case 0:
+                power_in_dbm=14;
+                break;
+            case 1:
+                power_in_dbm=23;
+                break;
+            case 2:
+                power_in_dbm=27;
+                break;
+            case 3:
+                power_in_dbm=29;
+                break;
+            }
+        }
+    }
+ return power_in_dbm;
+}
+
+
+/**
+ * Sends get settings command.
+ * */
+void AP_SmartAudio::request_settings()
+{
+    debug("%80s::request_settings()\t",TAG);
+    smartaudioFrame_t request;
+    uint8_t frame_size=smartaudioFrameGetSettings(&request);
+    packet command;
+    command.frame=request;
+    command.frame_size=frame_size;
+    requests_queue.push_force(command);
+}
+
+
+void AP_SmartAudio::set_operation_mode(uint8_t mode){
+     // take the first register from the output buffer
+   smartaudioSettings_t _current_state;
+
+    // peek from buffer
+   vtx_states_queue.peek(&_current_state,1);
+
+    debug("%80s::set_operation_mode(%02X)\t",TAG,mode);
+    // SPEC SAYS ONLY V2 SUPPORT SET MODE BUT THIS INCLUDES 2.1 ?
+    if (_current_state.version<SMARTAUDIO_SPEC_PROTOCOL_v2) {
+        debug("%s HW: %s",TAG,"Device can't change operation mode. Spec protocol not supported");
+        return;
+    }
+    smartaudioSettings_t settings;
+    settings.pitmodeInRangeActive=mode & 1<<0;
+    settings.pitmodeOutRangeActive=mode & 1<<1;
+    settings.pitModeRunning=mode & 1<<2;
+    settings.unlocked=(mode & 1<<3)!=0?1:0;
+
+    smartaudioFrame_t request;
+    uint8_t frame_size=smartaudioFrameSetOperationMode(&request,&settings);
+    packet command;
+    command.frame=request;
+    command.frame_size=frame_size;
+    requests_queue.push_force(command);
+
+}
+
+
+/**
+     * Sets the frecuency to transmit in the vtx.
+     * When isPitModeFreq active the frec will be set to be used when in pitmode (in range)
+     */
+void AP_SmartAudio::set_frequency(uint16_t frecuency,bool isPitModeFreq)
+{
+    debug("%80s::set_frequency(%d,%d)\t",TAG,frecuency,isPitModeFreq);
+    smartaudioFrame_t request;
+    uint8_t frame_size=smartaudioFrameSetFrequency(&request,frecuency,isPitModeFreq);
+    packet command;
+    command.frame=request;
+    command.frame_size=frame_size;
+    requests_queue.push_force(command);
```

Comment:
Do you really want push_force() here?

Reply:
The idea behind the push_force is to permit discard commands in case of floods from third party libraries. In theory this thirds parties, will be responsible to check that the command has been executed correctly calling to get_readings and checking if the expected change has been produced. 

I imagine the user interacting with the OSD, and changing options fast to setup the vtx, the idea is that last request ( so last options selected) where persited into the vtx.